src/test/java/org/opensearch/security/transport/SecurityInterceptorTests.java

```diff
@@ -201,17 +292,306 @@ public <T extends TransportResponse> void sendRequest(
         assertEquals(threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER), null);
     }
 
+    /**
+     * This test validates the behavior of the security interceptor when certain fields are null.
+     */
     @Test
-    public void testSendRequestDecorate() {
-        testSendRequestDecorate(Version.CURRENT);
+    public void testSendRequestDecorateWithNullFields() {
+        Version remoteNodeVersion = Version.CURRENT;
+        threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_USER, user);
+        boolean useJDKSerialization = remoteNodeVersion.before(ConfigConstants.FIRST_CUSTOM_SERIALIZATION_SUPPORTED_OS_VERSION);
+        ClusterName clusterName = ClusterName.DEFAULT;
+        when(clusterService.getClusterName()).thenReturn(clusterName);
+
+        MockTransport transport = new MockTransport();
+        TransportService transportService = transport.createTransportService(
+            Settings.EMPTY,
+            threadPool,
+            TransportService.NOOP_TRANSPORT_INTERCEPTOR,
+            boundTransportAddress -> clusterService.state().nodes().get(SecurityInterceptor.class.getSimpleName()),
+            null,
+            emptySet(),
+            NoopTracer.INSTANCE
+        );
+
+        // CS-SUPPRESS-SINGLE: RegexpSingleline Extensions manager used for creating a mock
+        OpenSearchSecurityPlugin.GuiceHolder guiceHolder = new OpenSearchSecurityPlugin.GuiceHolder(
+            mock(RepositoriesService.class),
+            transportService,
+            mock(IndicesService.class),
+            mock(PitService.class),
+            mock(ExtensionsManager.class)
+        );
+        // CS-ENFORCE-SINGLE
+
+        DiscoveryNode localNode = new DiscoveryNode("local-node1", new TransportAddress(localAddress, 1234), Version.CURRENT);
+        Connection connection1 = transportService.getConnection(localNode);
+
+        DiscoveryNode remoteNode = new DiscoveryNode("remote-node", new TransportAddress(localAddress, 6789), remoteNodeVersion);
+        Connection connection2 = transportService.getConnection(remoteNode);
+
+        // checking thread context inside sendRequestDecorate for connection1 with null local node -- we should serialize this
+        AsyncSender sender = new AsyncSender() {
+            @Override
+            public <T extends TransportResponse> void sendRequest(
+                Connection connection,
+                String action,
+                TransportRequest request,
+                TransportRequestOptions options,
+                TransportResponseHandler<T> handler
+            ) {
+                String serializedUserHeader = threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER);
+                assertEquals(serializedUserHeader, Base64Helper.serializeObject(user, useJDKSerialization));
+            }
+        };
```

Comment:
Lets make this more readable by creating a method that generates AsyncSender that takes in a function that could look like:

```java
final AsyncSender verifyUserInThreadContext = createAsyncSender((user) -> {
   final String serializedUserHeader = threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER);
   assertEquals(serializedUserHeader, Base64Helper.serializeObject(user, useJDKSerialization));
}); 
```

Also it might be worthwhile to make a helper function for what seems like an often repeated pattern `threadPool.getThreadContext().getHeader(ConfigConstants.OPENDISTRO_SECURITY_USER_HEADER)` -> getUserInThreadContext() to condense the tests

Reply:
Yeah sounds good; I was just keeping the original formatting as much as possible but I can swap it.